mmdet/models/detectors/dino.py

```diff
@@ -0,0 +1,535 @@
+
+        Returns:
+            dict: A dictionary of loss components
+        """
+        img_feats = self.extract_feat(batch_inputs)
+        head_inputs_dict = self.forward_transformer(
+            img_feats, batch_data_samples,
+            query_denoising=True)  # TODO: refine this  # noqa
+        losses = self.bbox_head.loss(
+            **head_inputs_dict, batch_data_samples=batch_data_samples)
+        return losses
+
+    def forward_transformer(
+            self,
+            img_feats: Tuple[Tensor],
+            batch_data_samples: OptSampleList = None,
+            query_denoising: bool = False
+    ) -> Dict:  # TODO: refine this  # noqa
+        """Forward process of Transformer.
+
+        The forward procedure of the transformer is defined as:
+        'pre_transformer' -> 'encoder' -> 'pre_decoder' -> 'decoder'
+        More details can be found at `TransformerDetector.forward_transformer`
+        in `mmdet/detector/base_detr.py`.
+        The difference is that the ground truth in `batch_data_samples` is
+        required for the `pre_decoder` to prepare the query of DINO.
+        Additionally, DINO inherits the `pre_transformer` method and the
+        `forward_encoder` method of DeformableDETR. More details about the
+        two methods can be found in `mmdet/detector/deformable_detr.py`.
+
+        Args:
+            img_feats (tuple[Tensor]): Tuple of feature maps from neck. Each
+                feature map has shape (bs, dim, H, W).
+            batch_data_samples (list[:obj:`DetDataSample`]): The batch
+                data samples. It usually includes information such
+                as `gt_instance` or `gt_panoptic_seg` or `gt_sem_seg`.
+                Defaults to None.
+            query_denoising (bool): Whether to calculate the denoising part.
+                Defaults to `False`.
+
+        Returns:
+            dict: The dictionary of bbox_head function inputs, which always
+            includes the `hidden_states` of the decoder output and may contain
+            `references` including the initial and intermediate references.
+        """
+        encoder_inputs_dict, decoder_inputs_dict = self.pre_transformer(
+            img_feats, batch_data_samples)
+
+        encoder_outputs_dict = self.forward_encoder(**encoder_inputs_dict)
+
+        tmp_dec_in, head_inputs_dict = self.pre_decoder(
+            **encoder_outputs_dict,
+            batch_data_samples=batch_data_samples,
+            query_denoising=query_denoising)  # TODO: refine this  # noqa
+        decoder_inputs_dict.update(tmp_dec_in)
+
+        decoder_outputs_dict = self.forward_decoder(**decoder_inputs_dict)
+        head_inputs_dict.update(decoder_outputs_dict)
+        return head_inputs_dict
+
+    def pre_decoder(
+        self,
+        memory: Tensor,
+        memory_mask: Tensor,
+        spatial_shapes: Tensor,
+        batch_data_samples: OptSampleList = None,
+        query_denoising: bool = False
```

Comment:
with_query_denoising or use_query_denosing to indicate it is a bool tensor.

Reply:
OK, I decided to use `self.training` as the indicator.